File: implementations/typescript/pact_eac.ts
type SemanticTypeName = string & { readonly semanticTypeName: unique symbol };
type SemanticOperatorName = string & { readonly semanticOperatorName: unique symbol };
type SemanticFieldPath = string & { readonly semanticFieldPath: unique symbol };
type SemanticEvidence = string & { readonly semanticEvidence: unique symbol };
type SemanticCheckIdentifier = string & { readonly semanticCheckIdentifier: unique symbol };
type SemanticValidationResult = boolean & { readonly semanticValidationResult: unique symbol };
type SemanticValue = number | string | readonly string[] | undefined;
type SemanticContext = Readonly<Record<string, Readonly<Record<string, SemanticValue>>>>;
type SemanticValidation = (value: SemanticValue) => SemanticValidationResult;
type SemanticPredicate = (left: SemanticValue, right: SemanticValue) => SemanticValidationResult;

const declareSemanticValidationResult = (result: boolean): SemanticValidationResult => result as SemanticValidationResult;
export const validateTypeIsInteger: SemanticValidation = value => declareSemanticValidationResult(Number.isInteger(value));
export const validateValueIsAtLeastMinimum = (minimum: number): SemanticValidation => value => declareSemanticValidationResult(typeof value === 'number' && value >= minimum);
export const validateValueIsAtMostMaximum = (maximum: number): SemanticValidation => value => declareSemanticValidationResult(typeof value === 'number' && value <= maximum);
export const composeAllValidationsMustPass = (...validations: readonly SemanticValidation[]): SemanticValidation => value => declareSemanticValidationResult(validations.every(validateCurrentRule => validateCurrentRule(value)));
export const validateIntegerRange = (minimum: number, maximum: number): SemanticValidation => composeAllValidationsMustPass(validateTypeIsInteger, validateValueIsAtLeastMinimum(minimum), validateValueIsAtMostMaximum(maximum));

export type ArithmeticExpression = Readonly<{ add?: readonly SemanticFieldPath[]; sub?: readonly SemanticFieldPath[] }>;
export type PolicyRule = Readonly<{ all_of?: readonly PolicyRule[]; any_of?: readonly PolicyRule[]; not?: PolicyRule; field?: SemanticFieldPath; expr?: ArithmeticExpression; op?: SemanticOperatorName; value?: SemanticValue }>;
export type PolicyCheck = Readonly<{ check_id: SemanticCheckIdentifier; rule: PolicyRule }>;
export type Mandate = Readonly<{ mandate_id: string; pdpp_version: string; principal_id: string; agent_id: string; issued_at: string; checks: readonly PolicyCheck[] }>;
export type PolicyEvaluationCheck = Readonly<{ check_id: SemanticCheckIdentifier; result: 'PASS' | 'FAIL'; evidence_ref: SemanticEvidence }>;

export const compareValuesAreEqual: SemanticPredicate = (left, right) => declareSemanticValidationResult(left === right);
export const compareLeftValueIsLessThanOrEqualToRightValue: SemanticPredicate = (left, right) => declareSemanticValidationResult(typeof left === 'number' && typeof right === 'number' && left <= right);
export const compareCollectionContainsValue: SemanticPredicate = (value, collection) => declareSemanticValidationResult(Array.isArray(collection) && typeof value === 'string' && collection.includes(value));
export const splitFieldPathIntoNamespaceAndKey = (path: SemanticFieldPath): readonly [string, string] => { const [namespace = '', key = ''] = String(path).split('.'); return [namespace, key]; };
export const resolveFieldPathFromSemanticContext = (context: SemanticContext, path: SemanticFieldPath): SemanticValue => { const [namespace, key] = splitFieldPathIntoNamespaceAndKey(path); return context[namespace]?.[key]; };
export const addResolvedFieldsFromSemanticContext = (context: SemanticContext, fields: readonly SemanticFieldPath[]): SemanticValue => fields.map(field => resolveFieldPathFromSemanticContext(context, field)).reduce((sum, value) => Number(sum) + Number(value), 0);
export const evaluateArithmeticExpression = (context: SemanticContext, expression: ArithmeticExpression): SemanticValue => expression.add ? addResolvedFieldsFromSemanticContext(context, expression.add) : undefined;
export const chooseSemanticPredicate = (operator: SemanticOperatorName): SemanticPredicate => ({ eq: compareValuesAreEqual, lte: compareLeftValueIsLessThanOrEqualToRightValue, in: compareCollectionContainsValue } as Record<string, SemanticPredicate>)[String(operator)] ?? compareValuesAreEqual;
export const evaluatePolicyRuleAgainstSemanticContext = (context: SemanticContext, rule: PolicyRule): readonly [SemanticValidationResult, SemanticEvidence] => { if (rule.all_of) return [declareSemanticValidationResult(rule.all_of.every(child => evaluatePolicyRuleAgainstSemanticContext(context, child)[0])), 'all_of children evaluated' as SemanticEvidence]; const actual = rule.expr ? evaluateArithmeticExpression(context, rule.expr) : resolveFieldPathFromSemanticContext(context, rule.field as SemanticFieldPath); return [chooseSemanticPredicate(rule.op as SemanticOperatorName)(actual, rule.value), 'semantic rule evaluated' as SemanticEvidence]; };
export const evaluateMandateAgainstTransactionAndInputs = (mandate: Mandate, transaction: Record<string, SemanticValue>, inputs: Record<string, SemanticValue>): readonly PolicyEvaluationCheck[] => { const context = { transaction, inputs }; return mandate.checks.map(check => { const [passed, evidence] = evaluatePolicyRuleAgainstSemanticContext(context, check.rule); return { check_id: check.check_id, result: passed ? 'PASS' : 'FAIL', evidence_ref: evidence }; }); };

File: implementations/go/pact_eac.go
package main

// PACT v0.2 expressed as Everything-as-Code in Go.
// Business values are loaded from ../../config.yml by the host application.

type SemanticTypeName string
type SemanticOperatorName string
type SemanticFieldPath string
type SemanticEvidence string
type SemanticCheckIdentifier string
type SemanticValidationResult bool

type SemanticValue any
type SemanticContext map[string]map[string]SemanticValue

type SemanticPredicate func(SemanticValue, SemanticValue) SemanticValidationResult
type SemanticValidation func(SemanticValue) SemanticValidationResult

type ArithmeticExpression struct{ Add []SemanticFieldPath `yaml:"add"`; Sub []SemanticFieldPath `yaml:"sub"` }
type PolicyRule struct { AllOf []PolicyRule `yaml:"all_of"`; AnyOf []PolicyRule `yaml:"any_of"`; Not *PolicyRule `yaml:"not"`; Field SemanticFieldPath `yaml:"field"`; Expr *ArithmeticExpression `yaml:"expr"`; Operator SemanticOperatorName `yaml:"op"`; Value SemanticValue `yaml:"value"` }
type PolicyCheck struct { CheckIdentifier SemanticCheckIdentifier `yaml:"check_id"`; Rule PolicyRule `yaml:"rule"` }
type Mandate struct { MandateIdentifier string `yaml:"mandate_id"`; PolicyVersion string `yaml:"pdpp_version"`; PrincipalIdentifier string `yaml:"principal_id"`; AgentIdentifier string `yaml:"agent_id"`; IssuedAt string `yaml:"issued_at"`; Checks []PolicyCheck `yaml:"checks"` }
type TransactionIntent map[string]SemanticValue
type PolicyInputs map[string]SemanticValue
type PolicyEvaluationCheck struct { CheckIdentifier SemanticCheckIdentifier `json:"check_id"`; Result string `json:"result"`; EvidenceReference SemanticEvidence `json:"evidence_ref"` }

func validateTypeIsInteger(value SemanticValue) SemanticValidationResult { _, valueIsInteger := value.(int); return SemanticValidationResult(valueIsInteger) }
func validateValueIsAtLeastMinimum(minimum SemanticValue) SemanticValidation { return func(value SemanticValue) SemanticValidationResult { return SemanticValidationResult(value.(int) >= minimum.(int)) } }
func validateValueIsAtMostMaximum(maximum SemanticValue) SemanticValidation { return func(value SemanticValue) SemanticValidationResult { return SemanticValidationResult(value.(int) <= maximum.(int)) } }
func composeAllValidationsMustPass(validations ...SemanticValidation) SemanticValidation { return func(value SemanticValue) SemanticValidationResult { for _, validateCurrentRule := range validations { if !validateCurrentRule(value) { return false } }; return true } }
func validateIntegerRange(minimum SemanticValue, maximum SemanticValue) SemanticValidation { return composeAllValidationsMustPass(validateTypeIsInteger, validateValueIsAtLeastMinimum(minimum), validateValueIsAtMostMaximum(maximum)) }

func compareValuesAreEqual(left SemanticValue, right SemanticValue) SemanticValidationResult { return SemanticValidationResult(left == right) }
func compareLeftValueIsLessThanOrEqualToRightValue(left SemanticValue, right SemanticValue) SemanticValidationResult { return SemanticValidationResult(left.(int) <= right.(int)) }
func compareCollectionContainsValue(value SemanticValue, collection SemanticValue) SemanticValidationResult { for _, candidate := range collection.([]SemanticValue) { if candidate == value { return true } }; return false }

func resolveFieldPathFromSemanticContext(context SemanticContext, path SemanticFieldPath) SemanticValue { namespace, key := splitFieldPathIntoNamespaceAndKey(path); return context[namespace][key] }
func splitFieldPathIntoNamespaceAndKey(path SemanticFieldPath) (string, string) { pathText := string(path); for index, character := range pathText { if character == '.' { return pathText[:index], pathText[index+1:] } }; return pathText, "" }
func addResolvedFieldsFromSemanticContext(context SemanticContext, fields []SemanticFieldPath) SemanticValue { total := 0; for _, field := range fields { total += resolveFieldPathFromSemanticContext(context, field).(int) }; return total }
func evaluateArithmeticExpression(context SemanticContext, expression ArithmeticExpression) SemanticValue { if len(expression.Add) > 0 { return addResolvedFieldsFromSemanticContext(context, expression.Add) }; return nil }
func chooseSemanticPredicate(operator SemanticOperatorName) SemanticPredicate { if operator == "eq" { return compareValuesAreEqual }; if operator == "lte" { return compareLeftValueIsLessThanOrEqualToRightValue }; if operator == "in" { return compareCollectionContainsValue }; return compareValuesAreEqual }
func evaluatePolicyRuleAgainstSemanticContext(context SemanticContext, rule PolicyRule) (SemanticValidationResult, SemanticEvidence) { if len(rule.AllOf) > 0 { for _, child := range rule.AllOf { passed, _ := evaluatePolicyRuleAgainstSemanticContext(context, child); if !passed { return false, "all_of child failed" } }; return true, "all_of children passed" }; actual := SemanticValue(nil); if rule.Expr != nil { actual = evaluateArithmeticExpression(context, *rule.Expr) } else { actual = resolveFieldPathFromSemanticContext(context, rule.Field) }; return chooseSemanticPredicate(rule.Operator)(actual, rule.Value), SemanticEvidence("semantic rule evaluated") }
func evaluateMandateAgainstTransactionAndInputs(mandate Mandate, transaction TransactionIntent, inputs PolicyInputs) []PolicyEvaluationCheck { context := SemanticContext{"transaction": map[string]SemanticValue(transaction), "inputs": map[string]SemanticValue(inputs)}; checks := []PolicyEvaluationCheck{}; for _, check := range mandate.Checks { passed, evidence := evaluatePolicyRuleAgainstSemanticContext(context, check.Rule); result := "FAIL"; if passed { result = "PASS" }; checks = append(checks, PolicyEvaluationCheck{check.CheckIdentifier, result, evidence}) }; return checks }
